infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/RoutingQueuesSnapshotReader.java

```diff
@@ -1,29 +1,288 @@
+                CorfuRuntime.getStreamID(streamTagFollowed));
+
+        // If the current opaque stream has data, read and send it
+        if (currentStreamHasNext()) {
+            msg = read(currentStreamInfo, syncRequestId);
+            if (msg != null) {
+                messages.add(msg);
+            }
+        } else {
+            // Wait for more data to be written by the client and send it once it is received
+            waitForData();
+            msg = read(currentStreamInfo, syncRequestId);
+            if (msg != null) {
+                messages.add(msg);
+            }
+        }
+        return new SnapshotReadMessage(messages, endMarkerReached);
+    }
+
+    // Waits for a default timeout period for data to get written for this destination on the Snapshot Sync queue
+    private void waitForData() {
+        // Create a task to wait for more data until DATA_WAIT_TIMEOUT_MS
+        StreamQueryTask streamQueryTask = new StreamQueryTask();
+        Future<Void> queryFuture = executorService.submit(streamQueryTask);
+
+        try {
+            queryFuture.get(DATA_WAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        } catch (TimeoutException e) {
+            throw new RuntimeException("Timed out waiting for data or end marker for Snapshot Sync");
+        } catch (Exception e) {
+            // Handle all other types of exceptions
+        }
+    }
+
+    private boolean isEndMarker(OpaqueEntry opaqueEntry) {
+        if (!opaqueEntry.getEntries().keySet().contains(endMarkerStreamId)) {
+            return false;
+        }
+
+        // There should be a single SMR update for the end marker
+        Preconditions.checkState(opaqueEntry.getEntries().get(endMarkerStreamId).size() == 1);
+        SMREntry smrEntry = opaqueEntry.getEntries().get(endMarkerStreamId).get(0);
+
+        // Deserialize the entry to get the marker
+        Object[] objs = smrEntry.getSMRArguments();
+
+        ByteBuf valueBuf = Unpooled.wrappedBuffer((byte[]) objs[1]);
+        CorfuRecord<Queue.RoutingTableSnapshotEndMarkerMsg, Message> record =
+            (CorfuRecord<Queue.RoutingTableSnapshotEndMarkerMsg, Message>)
+                replicationContext.getProtobufSerializer().deserialize(valueBuf, null);
+
+        Preconditions.checkState(Objects.equals(record.getPayload().getDestination(), session.getSinkClusterId()));
+
+        // TODO pankti: Pass the sync request id and return true only if the marker is for this snapshot sync
+        return true;
+    }
+
+    @Override
+    protected SMREntryList next(OpaqueStreamIterator stream) {
+        List<SMREntry> smrList = new ArrayList<>();
+        int currentMsgSize = 0;
+
+        try {
+            while (currentMsgSize < maxDataSizePerMsg) {
+                if (lastEntry != null) {
+
+                    // Only the entries from Routing Queue OR End Marker for Snapshot Sync must be found
+                    Preconditions.checkState(lastEntry.getEntries().keySet().size() == 1);
```

Comment:
I'm thinking we may not have control over end marker being in a separate transaction.
Since you have the end marker as a separate stream, it doesn't matter if its in a separate txn or not right?

Reply:
if the end marker gets written in the same transaction, the keySet size will be > 1 and this check will fail.

I was working under the assumption that the client will send an end marker to LR through the client APIs, where we can control which transaction it gets written to